api_files/generator.py

```diff
@@ -0,0 +1,191 @@
+'''
+Takes in the api.yaml file from swagger and generates a new one that 
+has only the required properties in it for validation.
+Also generates a state_changes.yaml which defines which properties 
+are allowed in Scenes/State
+'''
+
+import yaml, copy
+from decouple import config
+from logger import Logger, LogLevel
+
+SWAGGER_YAML = config('SWAGGER_YAML')
+NEW_API = config('API_YAML')
+STATE_CHANGES = config('STATE_YAML')
+
+class ApiGenerator:
+    logger = Logger("apiGenerator")
+    api_file = None
+    api_yaml = None
+    new_api_file = None
+    new_state_file = None
+
+    def __init__(self):
+        '''
+        Load in the files
+        '''
+        self.new_api_file = open(NEW_API, 'w')
+        self.new_state_file = open(STATE_CHANGES, 'w')
+        try:
+            self.api_file = open(SWAGGER_YAML)
+            self.api_yaml = yaml.load(self.api_file, Loader=yaml.CLoader)
+        except Exception as e:
+            self.logger.log(LogLevel.ERROR, "Error while loading in api yaml. Please check the .env to make sure the location is correct and try again.\n\n" + str(e) + "\n")
+
+    def __del__(self):
+        '''
+        Basic cleanup: closing the file loaded in on close.
+        '''
+        self.logger.log(LogLevel.DEBUG, "Program closing...")
+        if (self.new_api_file):
+            self.new_api_file.close()
+        if (self.api_file):
+            self.api_file.close()
+        if (self.new_state_file):
+            self.new_state_file.close()
+    
+    def update_general_api(self):
+        '''
+        Updates properties in the api.yaml file that don't match with
+        what we expect during TA1's validation.
+        '''
+        new_api = copy.deepcopy(self.api_yaml)
+        # validator requires Scenario/Scenes
+        required_scenario = new_api['components']['schemas']['Scenario']['required']
+        required_scenario.append('scenes')
+
+        # validator requires Action/probe_id and Action/choice
+        required_actions = new_api['components']['schemas']['Action']['required']
+        required_actions.append('probe_id')
+        required_actions.append('choice')
+        
+        # validator requires all vitals properties to be specified
+        required_vitals = new_api['components']['schemas']['Vitals']['required'] if 'required' in new_api['components']['schemas']['Vitals'] else []
+        required_vitals.append('conscious')
+        required_vitals.append('avpu')
+        required_vitals.append('mental_status')
+        required_vitals.append('breathing')
+        required_vitals.append('hrpmin')
+        required_vitals.append('Spo2')
+        new_api['components']['schemas']['Vitals']['required'] = required_vitals
+
+        # injury status enum should only include hidden, discoverable, or visible
+        new_api['components']['schemas']['InjuryStatusEnum']['enum'] = ['hidden', 'discoverable', 'visible']
+
```

Comment:
Can we add a new rule I just thought of:
- `restricted_actions` cannot include `TAG_CHARACTER` and `END_SCENARIO`.  I don't want them to have a situation where they've set `end_scenario_allowed` to true but `restricted_actions` to include `END_SCENARIO`.

Reply:
So do you want this to always be true? Or do you want me to check if they set end_scenario_allowed to true and also restricted the actions?

Reply:
It should always be true for `restricted_actions`.  It's limiting the allowable controlled vocab for this field.  If that makes it special-case enough that it needs to be in another ticket, then that's a-ok.  I don't think you need to cross-reference with `end_scenario_allowed`.

Reply:
I just threw some code in that creates a new enum for restricted_actions based on actionTypeEnum, should work now!